monitoring/src/minting-monitor.ts

```diff
@@ -1,11 +1,181 @@
+// person to investigate the problem.
+const OptimisticMintingRequestedForUndeterminedBtcTx = (
+  chainEvent: OptimisticMintingRequestedChainEvent,
+  btcClientResponse: string
+): SystemEvent => ({
+  title: "Optimistic minting requested for undetermined Bitcoin transaction",
+  type: SystemEventType.Critical,
+  data: {
+    minter: `0x${chainEvent.minter.identifierHex}`,
+    depositKey: chainEvent.depositKey.toPrefixedString(),
+    depositor: `0x${chainEvent.depositor.identifierHex}`,
+    amountSat: chainEvent.amount.div(satoshiMultiplier).toString(),
+    btcFundingTxHash: chainEvent.fundingTxHash.toString(),
+    btcFundingOutputIndex: chainEvent.fundingOutputIndex.toString(),
+    btcClientResponse,
+    ethRequestTxHash: chainEvent.transactionHash.toPrefixedString(),
+  },
+  block: chainEvent.blockNumber,
+})
 
 export class MintingMonitor implements SystemEventMonitor {
+  private bridge: Bridge
+
+  private tbtcVault: TBTCVault
+
+  private btcClient: BitcoinClient
+
+  constructor(bridge: Bridge, tbtcVault: TBTCVault, btcClient: BitcoinClient) {
+    this.bridge = bridge
+    this.tbtcVault = tbtcVault
+    this.btcClient = btcClient
+  }
+
   async check(fromBlock: number, toBlock: number): Promise<SystemEvent[]> {
-    // TODO: Implement the minting monitor.
-    console.log(
-      `checked stub minting monitor from block ${fromBlock} to block ${toBlock}`
+    // eslint-disable-next-line no-console
+    console.log("running minting monitor check")
+
+    const systemEvents: SystemEvent[] = []
+
+    systemEvents.push(...(await this.checkCancelledEvents(fromBlock, toBlock)))
+
+    systemEvents.push(...(await this.checkRequestedEvents(fromBlock, toBlock)))
+
+    // eslint-disable-next-line no-console
+    console.log("completed minting monitor check")
+
+    return systemEvents
+  }
+
+  private async checkCancelledEvents(fromBlock: number, toBlock: number) {
+    const chainEvents =
+      await this.tbtcVault.getOptimisticMintingCancelledEvents({
+        fromBlock,
+        toBlock,
+      })
+
+    return chainEvents.map(OptimisticMintingCancelled)
+  }
+
+  private async checkRequestedEvents(fromBlock: number, toBlock: number) {
+    const chainEvents =
+      await this.tbtcVault.getOptimisticMintingRequestedEvents({
+        fromBlock,
+        toBlock,
+      })
+
+    const confirmations = await Promise.allSettled(
+      chainEvents.map((ce) =>
+        this.btcClient.getTransactionConfirmations(ce.fundingTxHash)
+      )
     )
-    return []
+
+    const systemEvents: SystemEvent[] = []
+
+    chainEvents.forEach((ce, index) => {
+      const confirmation = confirmations[index]
+
+      switch (confirmation.status) {
+        case "fulfilled": {
+          const actualConfirmations = confirmation.value
+          const requiredConfirmations = this.requiredConfirmations(
+            ce.amount.div(satoshiMultiplier)
+          )
+
+          if (actualConfirmations < requiredConfirmations) {
```

Comment:
Note that this logic is valid for the time being but may require updates.

Right now, @nkuba is working on an update to Minter to request optimistic minting earlier and Guardian to cancel if conditions are not fulfilled by the end of the optimistic minting delay.

Reply:
Yep, I'm aware. We must remember that the monitoring must evolve along with the monitored components and adjust things when the time comes.